### PR TITLE
Allow using a CSV file to define currents for hydrodynamic system

### PR DESCRIPTION
**Pull request by arjo129**
🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸

# 🎉 New feature

Closes #<NUMBER>

## Summary
This PR adds support for looking up hydrodynamic currents from a data file via the new Environment component. This allows users to vary currents throughout the water column.

Adds the tags:
```
        <lookup_current_x>current_x</lookup_current_x>
        <lookup_current_y>current_y</lookup_current_y>
        <lookup_current_z>current_z</lookup_current_z>
```
When present, they use the tag to look up current from the Environment component. If one is not present, that column defaults to zero.

Signed-off-by: Arjo Chakravarty <arjo@openrobotics.org>

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: src/systems/hydrodynamics/Hydrodynamics.cc [hunk 1]
```diff
@@ -78,14 +79,121 @@ class gz::sim::systems::HydrodynamicsPrivateData
...
   public: Eigen::VectorXd prevStateDot;
 
-  /// Link entity
+  /// \brief Use current table if true
+  public: bool useCurrentTable {false};
+
+  /// \brief Current table xComponent
+  public: std::string axisComponents[3];
+
+  public: std::shared_ptr<gz::sim::v7::components::EnvironmentalData>
```
Comment (hidmic) [inline src/systems/hydrodynamics/Hydrodynamics.cc:91]: @arjo129 nit: do we need to specify the version namespace? It should be inlined.

File: src/systems/hydrodynamics/Hydrodynamics.cc [hunk 2]
```diff
@@ -78,14 +79,121 @@ class gz::sim::systems::HydrodynamicsPrivateData
...
+
+  /// Retrieve the currernt from the data table
+  public: math::Vector3d GetCurrentFromEnvironment(
+    const EntityComponentManager &_ecm,
+    const std::chrono::steady_clock::duration &_currTime,
+    math::Vector3d _position)
+  {
+    math::Vector3d current(0, 0, 0);
+
+    if (!this->ready)
```
Comment (hidmic) [inline src/systems/hydrodynamics/Hydrodynamics.cc:155]: @arjo129 nit: isn't `ready` equivalent to having (1) a data frame and (b) at least one session for some axis? Do we need the (redundant) boolean flag?

**Review by hidmic — COMMENTED** [review]
First pass. Code formatting is a bit odd but otherwise it's looking good.

File: src/systems/hydrodynamics/Hydrodynamics.cc [hunk 3]
```diff
@@ -78,14 +79,121 @@ class gz::sim::systems::HydrodynamicsPrivateData
...
+        {
+          auto data = _environmental_data->Data();
+          if (!data->frame.Has(this->axisComponents[i]))
+          {
+            gzwarn << "Environmental sensor could not find field "
+              << this->axisComponents[i] << "\n";
+            continue;
+          }
+
+          this->gridField[i] = data;
```
[the remark this answers is not in the document]
Comment (arjo129) [reply src/systems/hydrodynamics/Hydrodynamics.cc]: baba5f2

File: test/integration/hydrodynamics.cc [hunk 4]
```diff
@@ -50,6 +50,8 @@ class HydrodynamicsTest : public InternalFixture<::testing::Test>
   /// \param[in] _drag_coeff Body drag coefficient
   public: std::vector<math::Vector3d> TestWorld(const std::string &_world,
    const std::string &_namespace);
+
+  public: math::Vector3d defaultForce{0, 0, 10.0};
```
[the remark this answers is not in the document]
Comment (arjo129) [reply test/integration/hydrodynamics.cc:54]: No. My intent was to vary this force, I messed up and didnt commit those changes. See baba5f2

File: src/systems/hydrodynamics/Hydrodynamics.hh [hunk 5]
```diff
@@ -94,6 +96,15 @@ namespace systems
...
   ///     To be deprecated in Garden.
   ///
+  /// ### Loading external currents
+  /// One can use the EnvironmentPreload system to preload currents into the
+  /// plugin using data files. To use the data you may give CSV column names by
+  /// using `lookup_current_*` tags listed below:
+  ///   * <lookup_current_x> - X axis to use for lookup current
+  ///   * <lookup_current_y> - Y axis to use for lookup current
+  ///   * <lookup_current_z> - Z axis to use for lookup current
+  /// This would provide
```
[the remark this answers is not in the document]
Comment (arjo129) [reply src/systems/hydrodynamics/Hydrodynamics.hh]: [dc05823](https://github.com/gazebosim/gz-sim/pull/1839/commits/dc058234986d1b9616830cd6d5602ede0d7fb4db)

File: src/systems/hydrodynamics/Hydrodynamics.cc [hunk 6]
```diff
@@ -78,14 +79,121 @@ class gz::sim::systems::HydrodynamicsPrivateData
...
   /// \brief Ocean current callback
   public: void UpdateCurrent(const msgs::Vector3d &_msg);
 
+  /// \brief Set the current table
+  public: void SetCurrentTable(
+    const EntityComponentManager &_ecm,
+    const std::chrono::steady_clock::duration &_currTime)
+  {
+    _ecm.EachNew<components::Environment>([&](const Entity &/*_entity*/,
+    const components::Environment *_environmental_data)->bool
```
[the remark this answers is not in the document]
Comment (arjo129) [reply src/systems/hydrodynamics/Hydrodynamics.cc]: [dc05823](https://github.com/gazebosim/gz-sim/pull/1839/commits/dc058234986d1b9616830cd6d5602ede0d7fb4db)

File: src/systems/hydrodynamics/Hydrodynamics.cc [hunk 7]
```diff
@@ -78,14 +79,138 @@ class gz::sim::systems::HydrodynamicsPrivateData
...
+  public: math::Vector3d GetCurrentFromEnvironment(
+    const EntityComponentManager &_ecm,
+    const std::chrono::steady_clock::duration &_currTime,
+    math::Vector3d _position)
+  {
+    math::Vector3d current(0, 0, 0);
+
+    if (!this->gridField ||
+      !(this->session[0].has_value() ||
+        this->session[1].has_value() || this->session[1].has_value()))
```
Comment (hidmic) [inline src/systems/hydrodynamics/Hydrodynamics.cc:163]: @arjo129 I think it should be
```suggestion
        this->session[1].has_value() || this->session[2].has_value()))
```

File: src/systems/hydrodynamics/Hydrodynamics.cc [hunk 8]
```diff
@@ -78,14 +79,138 @@ class gz::sim::systems::HydrodynamicsPrivateData
...
+        auto position = getGridFieldCoordinates(
+          _ecm, _position, this->gridField);
+
+        if (!position.has_value())
+        {
+          gzerr << "Coordinate conversion failed" << std::endl;
+          continue;
+        }
+
+        if (!this->session[i].has_value())
```
Comment (hidmic) [inline src/systems/hydrodynamics/Hydrodynamics.cc:189]: @arjo129 nit: consider having this check before coordinate conversion

File: test/integration/hydrodynamics.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (hidmic) [reply test/integration/hydrodynamics.cc:54]: It still seems unchanged?

File: test/worlds/hydrodynamics.sdf.in [hunk 9]
```diff
@@ -11,9 +11,24 @@
     <!-- prevent sinking -->
     <gravity>0 0 0</gravity>
 
+    <!-- Current Data -->
+    <plugin
+      filename="gz-sim-environment-preload-system"
+      name="gz::sim::systems::EnvironmentPreload">
+      <data>@CMAKE_SOURCE_DIR@/test/worlds/hydrodynamic_currents.csv</data>
```
Comment (hidmic) [inline test/worlds/hydrodynamics.sdf.in:18]: @arjo129 fyi: if both SDF and CSV files are sitting next to each other, you can use a relative path here.

**Review by hidmic — COMMENTED** [review]
Second pass. Looking good!

File: test/integration/hydrodynamics.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (arjo129) [reply test/integration/hydrodynamics.cc:54]: https://github.com/gazebosim/gz-sim/blob/dc058234986d1b9616830cd6d5602ede0d7fb4db/test/integration/hydrodynamics.cc#L183

File: src/systems/hydrodynamics/Hydrodynamics.cc [hunk 10]
```diff
@@ -78,14 +79,136 @@ class gz::sim::systems::HydrodynamicsPrivateData
...
   public: Entity linkEntity;
 
   /// \brief Ocean current callback
   public: void UpdateCurrent(const msgs::Vector3d &_msg);
 
+  /////////////////////////////////////////////////
+  /// \brief Set the current table
+  /// \param[in] _ecm - The Entity Component Manager
+  /// \param[in] _currTime - The current time
+  public: void SetCurrentTable(
```
Comment (hidmic) [inline src/systems/hydrodynamics/Hydrodynamics.cc:106]: @arjo129 nit^N: current could be understood as present. Consider qualifying it: ocean current, water current, you name it.

**Review by hidmic — APPROVED** [review]

**Pull request merged**